File: packages/qnsi/src/crypto-inventory.ts
/**
 * QNSI Crypto-Inventory (CBOM) - asset catalogue, discovery runs,
 * deprecation policies, PQC migration readiness. Wraps
 * `apps/crypto-inventory-service` (`/crypto/v1`).
 */

import type { Internal, RequestOptions } from "./_internal.js";

const PATH_PREFIX = "/proxy/crypto/v1";

export interface DiscoverAssetsRequest {
	readonly targets?: readonly string[];
	readonly modes?: readonly string[];
	readonly options?: Record<string, unknown>;
}

export class CryptoInventoryClient {
	constructor(private readonly internal: Internal) {}

	listAssets(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/assets`, undefined, { query });
	}

	getAsset(assetId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/assets/${assetId}`);
	}

	/** @param _tenantId deprecated and ignored - the tenant is derived from the API key. */
	getAssetStats(_tenantId?: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/assets/stats`);
	}

	discoverAssets(req: DiscoverAssetsRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/assets/discover`, req, opts);
	}

	/**
	 * Basic readiness summary (`/readiness`). For the scored PQC-migration
	 * attestation prefer {@link getPqcReadinessScore}.
	 * @param _tenantId deprecated and ignored - the tenant is derived from the API key.
	 */
	getReadinessScore(_tenantId?: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/readiness`);
	}

	/**
	 * PQC migration-readiness score for the tenant (0-100 with category
	 * breakdown) - the score behind a "PQC-ready" attestation.
	 */
	getPqcReadinessScore() {
		return this.internal.request("GET", `${PATH_PREFIX}/pqc-readiness/score`);
	}

	/** Actionable PQC-migration recommendations for the tenant. */
	getPqcReadinessRecommendations() {
		return this.internal.request("GET", `${PATH_PREFIX}/pqc-readiness/recommendations`);
	}

	/**
	 * Aggregated Cryptographic Bill of Materials (CycloneDX 1.5) for the tenant -
	 * the machine-verifiable inventory behind the conformance/attestation claim.
	 */
	getCbom(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/cbom`, undefined, { query });
	}

	/** List discovery runs for the tenant. */
	listDiscoveryRuns(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/discovery/runs`, undefined, { query });
	}

	/**
	 * Import a batch of cryptographic assets (scan output / inventory upload)
	 * into the tenant's CBOM.
	 */
	importAssets(body: Record<string, unknown>, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/assets/import`, body, opts);
	}
}

File: sdks/rust/qnsi/src/crypto_inventory.rs
//! QNSI Crypto-Inventory (CBOM) - asset catalogue, discovery runs,
//! deprecation policies, and PQC migration readiness.

use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::sync::Arc;

use crate::activation::Activation;
use crate::errors::Error;
use crate::service_client::ServiceClient;

const PATH_PREFIX: &str = "/proxy/crypto/v1";

#[derive(Clone)]
pub struct Client {
    sc: ServiceClient,
}

impl Client {
    pub fn new(activation: Arc<Activation>, http: reqwest::Client) -> Self {
        Self { sc: ServiceClient::new(activation, http, PATH_PREFIX) }
    }

    pub async fn list_assets(&self, query: &[(&str, String)]) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/assets", None, Some(query), None).await
    }

    pub async fn get_asset(&self, asset_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, &format!("/assets/{}", asset_id), None, None, None).await
    }

    pub async fn get_asset_stats(&self, _tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/assets/stats", None, None, None).await
    }

    pub async fn discover_assets(&self, req: DiscoverAssetsRequest, idempotency_key: Option<&str>) -> Result<Value, Error> {
        self.sc.request(Method::POST, "/assets/discover", Some(&req), None, idempotency_key).await
    }

    pub async fn get_readiness_score(&self, _tenant_id: &str) -> Result<Value, Error> {
        self.sc.request::<()>(Method::GET, "/readiness", None, None, None).await
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoverAssetsRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub targets: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modes: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<serde_json::Map<String, Value>>,
}
